Give Programs Access to Code in Subdirectories

everythingfunctional opened the pull request:
Programs can now use modules in subdirectories from their location. I.e. a program `app/main.f90` can use a module `app/sub_dir/mod.f90`. This applies to tests and examples as well.

Fix #408

everythingfunctional commented:
Sorry about the noisy change set. My editor is set to automatically remove trailing whitespace.

awvwgk commented on fpm/src/fpm_targets.f90:
```diff
@@ -409,7 +409,7 @@ function find_module_dependency(targets,module_name,include_dir) result(target_p
                 case default
                     if (present(include_dir)) then
-                        if (dirname(targets(k)%ptr%source%file_name) == include_dir) then
+                        if (index(dirname(targets(k)%ptr%source%file_name), include_dir) > 0) then ! source file is within the include_dir or a subdirectory
```
Should we check if index is 1 rather than larger zero? A match at position 10 could be a false positive.

everythingfunctional replied:
I suppose that's true. Good catch.

LKedward commented on fpm/test/fpm_test/test_module_dependencies.f90:
```diff
@@ -356,19 +356,19 @@ subroutine test_invalid_library_use(error)
     end subroutine test_invalid_library_use
 
 
     !> Check program using a non-library module in a different directory
```
```suggestion
    !> Check program using a non-library module in a subdirectory
```

LKedward commented on fpm/src/fpm_targets.f90:
```diff
@@ -314,22 +314,22 @@ subroutine add_dependency(target, dependency)
 !> - Executable sources (`FPM_SCOPE_APP`,`FPM_SCOPE_TEST`) may use
 !>   library modules (including dependencies) as well as any modules
-!>   corresponding to source files __in the same directory__ as the 
+!>   corresponding to source files __in the same directory__ as the
 !>   executable source.
```
```suggestion
!>   corresponding to source files in the same directory or a 
!>   subdirectory of the executable source file.
```

LKedward reviewed (approved):
Many thanks. This all looks good to me. My only request is are you able to add a failing test for a program using a module in a parent directory, to cover all conditions.

everythingfunctional commented:
I've added the test and updated the doc string as suggested. Thanks for the great reviews guys. I'll give you guys a few hours in case you want to double check the changes and merge later today.
